src/simularium/WebsocketClient.ts

```diff
@@ -0,0 +1,317 @@
+import jsLogger from "js-logger";
+import { ILogger } from "js-logger";
+import { FrontEndError, ErrorLevel } from "./FrontEndError";
+
+export interface NetMessage {
+    connId: string;
+    msgType: number;
+    fileName: string;
+}
+// TODO: proposed new NetMessage data type:
+// This factors the raw data structure away from the networking and transmission info.
+// This allows the data structure to make a bit more sense with respect to typescript typing,
+// and also for raw file drag n drop it doesn't need connection info or msgtype.
+// interface NetMessage {
+//     connId: string; // unique connection to server
+//     msgType: number; // identifies the data structure of the message
+//     fileName: string; // identifies the trajectory this connection is dealing with
+//     payload: Object; // the JS object with the message data itself
+// }
+
+export interface MessageEventLike {
+    data: string;
+}
+
+// these have been set to correspond to backend values
+export const enum NetMessageEnum {
+    ID_UNDEFINED_WEB_REQUEST = 0,
+    ID_VIS_DATA_ARRIVE = 1,
+    ID_VIS_DATA_REQUEST = 2,
+    ID_VIS_DATA_FINISH = 3,
+    ID_VIS_DATA_PAUSE = 4,
+    ID_VIS_DATA_RESUME = 5,
+    ID_VIS_DATA_ABORT = 6,
+    ID_UPDATE_TIME_STEP = 7,
+    ID_UPDATE_RATE_PARAM = 8,
+    ID_MODEL_DEFINITION = 9,
+    ID_HEARTBEAT_PING = 10,
+    ID_HEARTBEAT_PONG = 11,
+    ID_TRAJECTORY_FILE_INFO = 12,
+    ID_GOTO_SIMULATION_TIME = 13,
+    ID_INIT_TRAJECTORY_FILE = 14,
+    ID_UPDATE_SIMULATION_STATE = 15,
+    ID_CONVERT_TRAJECTORY_FILE = 16,
+    ID_CONVERTED_TRAJECTORY = 17,
+    ID_CALCULATE_METRICS = 18,
+    ID_METRICS_PLOT_DATA = 19,
+    // insert new values here before LENGTH
+    LENGTH,
+}
+
+export const CONNECTION_SUCCESS_MSG = "Remote sim successfully started";
+export const CONNECTION_FAIL_MSG =
+    "Failed to connect to server; try reloading. If the problem persists, there may be a problem with your connection speed or the server might be too busy.";
+
+export interface NetConnectionParams {
+    serverIp?: string;
+    serverPort?: number;
+}
+
+export class WebsocketClient {
+    private webSocket: WebSocket | null;
+    private serverIp: string;
+    private serverPort: number;
+    public connectionTimeWaited: number;
+    public connectionRetries: number;
+    protected jsonMessageHandlers: Map<NetMessageEnum, (NetMessage) => void>;
+    protected binaryMessageHandlers: Map<
+        NetMessageEnum,
+        (MessageEvent) => void | ((MessageEventLike) => void)
```

Comment:
Instead of having `MessageEvent | MessageEventLike` in many different places, I think what would look cleaner here is to define a type called `MessageEventLike` that is `MessageEvent | {data:string}`, and then use MessageEventLike everywhere.
You could name the {data:string} as SimulariumEvent or something like that.